### DISCUSSION TITLE
"Keying" cache volumes

### DISCUSSION THREAD

**Discussion by rgilton** [dagger/dagger#9252] [Help]
We have several projects using the same dagger CI runner.  So if one project creates a cache volume called `pypi`, then all of the other projects would also be able to share that cache.  Ideally we'd like to isolate the caches of different projects.

One approach is to extend the cache volume's name ("key" it) with some project-specific string (e.g. `pypi-foo` for project `foo`, and `pypi-bar` for project `bar`), but this seems like it could be pretty error-prone.  If we fail to extend on of the cache names, then we may end up with unintended contamination between projects.

Furthermore, we also would ideally use separate caches for our "release" pipelines from our merge-request pipelines.  A similar approach could be used here too (`pypi-foo-release`, etc).

What would be great is some way within Dagger to add a key to _all_ caches used within a pipeline to allow isolation of projects. For example, by setting a cli arg to `dagger run` (e.g. `--cache-key`), or setting an environment variable (e.g. `DAGGER_CACHE_KEY`).  Does such functionality already exist?  If not, can we get it :laughing:?

**Comment by jedevc** [comment]
I think this is similar to the issue described in https://github.com/dagger/dagger/issues/7211:

> Cache volumes exist in the engine's global namespace: if modules A and B each create a cache volume with id foo, they will share read and write access to the same volume, as long as they are run on the same engine. This allows one module to corrupt the data of another module, either accidentally or maliciously.

Modules are the natural way to namespace this - these already have pretty much unique identifiers, so namespacing cache volumes along these lines make the most sense.

This is actually already in-progress in https://github.com/dagger/dagger/pull/8724 - hopefully that would solve your issue? :tada:

**Reply by rgilton** [reply]
Hmm, it doesn't sound like it really solves the issue.  If there's a module used between two projects then they'd share the cache, and they would no longer be isolated.  It also wouldn't cover isolating merge requests from release pipelines.

It doesn't actually cover our use case either, as we haven't "bought into the daggerverse" either -- we've been finding it much simpler to just use the dagger python API directly, rather than introducing an extra layer of packaging on top of everything.

**Reply by rgilton** [reply]
I can see #8724 adds a "namespace" parameter to each volume.  I was hoping for a solution that allows us to globally append some string to all namespaces in a process, or something like that.
